feat: generate TaskChecks by the corresponding SchemaReviewPolicy

rebelice opened the pull request:
Schema review policy consists of a list of schema review rules.
 There is such a logical mapping in bytebase backend:
   1. One `schema review policy` maps a `TaskCheckRun`.
   2. Each `schema reivew rule type` maps an `advisor.Type`.
   3. Each `[db.Type][AdvisorType]` maps an `advisor`.

## How to add a schema review rule:
   1. Implement an advisor.(plugin/xxx)
   2. Register this advisor in `map[db.Type][AdvisorType]`.(plugin/advisor.go)
   3. Map `SchemaReviewRuleType` to `advisor.Type`.(server/schema_system.go:getAdvisorTypeByRule())

https://www.loom.com/share/fe9204a0e3f2413c90830b1d5290f6d2

tianzhou commented on api/task_check_run.go:
```diff
@@ -92,6 +92,12 @@ type TaskCheckDatabaseStatementAdvisePayload struct {
 	Collation string  `json:"collation,omitempty"`
+
+	// Schema review rule special fields.
+	// Level is the alert level for this schem review rule level.
```
```suggestion
	// Level is the alert level for this schema review rule level.
```

tianzhou commented on plugin/advisor/advisor.go:
```diff
@@ -62,6 +75,13 @@ type Context struct {
 	Collation string
+
+	// Schema review rule special fields.
+	// Level is the alert level for this schem review rule level.
```
```suggestion
	// Level is the alert level for this schema review rule level.
```

tianzhou commented on server/task_check_scheduler.go:
```diff
@@ -347,6 +365,43 @@ func (s *TaskCheckScheduler) ScheduleCheckIfNeeded(ctx context.Context, task *ap
+		if rule.Level == api.SchemaRuleLevelDisabled {
+			continue
+		}
+		taskCheckType, payload, err := getTaskCheckTypeAndPayloadByRule(rule, basePayload)
```
I am thinking of only having a single task check type for all schema review rules.

Otherwise. there are just too many task checks (and each will surface on the UI).

tianzhou commented on server/task_check_executor_statement_advisor.go:
```diff
@@ -52,6 +52,8 @@ func (exec *TaskCheckStatementAdvisorExecutor) Run(ctx context.Context, server *
 			Logger:    exec.l,
 			Charset:   payload.Charset,
 			Collation: payload.Collation,
+			Level:     payload.Level,
```
curious why we create a separate Level field instead of putting it into the Payload.

tianzhou reviewed (changes requested):
Also need to add frontend code otherwise the checker bar name is empty.

https://sourcegraph.com/search?q=context:%40bytebase+bb.task-check.database.statement.compatibility&patternType=literal

rebelice commented:
I'll refactor this PR to follow this comment https://github.com/bytebase/bytebase/pull/1169#discussion_r858510473

ecmadao reviewed (approved):
LGTM

tianzhou commented on server/task_check_scheduler.go:
```diff
@@ -333,6 +333,24 @@ func (s *TaskCheckScheduler) ScheduleCheckIfNeeded(ctx context.Context, task *ap
 			}
 		}
 
+		if s.server.feature(api.FeatureSchemaReviewPolicy) {
```
If we do the refactor in the followup PR, then the change here needs to be disabled. Otherwise, every schema change task will trigger the check.

boojack commented on frontend/src/locales/en.yml:
```diff
@@ -219,6 +219,7 @@ task:
     compatibility: Compatibility
     connection: Connection
     migration-schema: Migration schema
+    schema-review: Schema review
```
Also needs to add the translation to `./zh-CN.yml`

tianzhou replied:
Let's call this SQL review since it also applies to Data Change. Also this surfaces to the end-user, though our underlying model is called statement instead of SQL, the current specific use case is SQL, so let's use "SQL".

This is similar to other places that we surface "SQL" instead of "Statement" in the UI.

rebelice commented:
I have refactored this PR and addressed all comments.
I re-describe the new mapping and re-record the demo in https://github.com/bytebase/bytebase/pull/1169#issue-1215589577.
PTAL @tianzhou @d-bytebase @ecmadao @boojack

LiuJi-Jim reviewed (approved):
LGTM

tianzhou commented on api/task_check_run.go:
```diff
@@ -71,8 +71,8 @@ const (
-	// TaskCheckDatabaseStatementRequireWhere is the task check type for the WHRER clause requirement for UPDATE/DELETE.
-	TaskCheckDatabaseStatementRequireWhere TaskCheckType = "bb.task-check.database.statement.where.require"
+	// TaskCheckDatabaseStatementSchemaReview is the task check type for schema system review policy.
+	TaskCheckDatabaseStatementSchemaReview TaskCheckType = "bb.task-check.database.statement.schema-review"
```
TaskCheckDatabaseStatementAdvise?

schema review is too specific. e.g. Data change is not schema and it requires enforcing the policy.

Advise is a bit general, while it's at least more accurate than schema review

tianzhou commented on plugin/advisor/mysql/mysql_where_requirement_advisor.go:
```diff
@@ -48,14 +51,15 @@ func (adv *WhereRequirementAdvisor) Check(ctx advisor.Context, statement string)
 			Code:    common.Ok,
 			Title:   "OK",
-			Content: "",
+			Content: "Pass rule: statements require 'WHERE' clause",
```
I would just use "". People won't be that interested in the happy case.

tianzhou commented on server/task_check_executor_statement_advisor.go:
```diff
@@ -41,8 +37,16 @@ func (exec *TaskCheckStatementAdvisorExecutor) Run(ctx context.Context, server *
-		advisorType = advisor.MySQLWhereRequirement
+	case api.TaskCheckDatabaseStatementSchemaReview:
+		if !server.feature(api.FeatureSchemaReviewPolicy) {
+			return []api.TaskCheckResult{}, common.Errorf(common.NotAuthorized, fmt.Errorf(api.FeatureBackwardCompatibility.AccessErrorMessage()))
```
copy-paste error

tianzhou commented on server/task_check_executor_statement_advisor.go:
```diff
@@ -59,7 +63,56 @@ func (exec *TaskCheckStatementAdvisorExecutor) Run(ctx context.Context, server *
+		}
+		advisorType, err := getAdvisorTypeByRule(rule.Type)
+		if err != nil {
+			exec.l.Debug("rule not support", zap.Error(err))
```
```suggestion
			exec.l.Debug("not supported rule", zap.Error(err))
```

ecmadao commented on server/task_check_executor_statement_advisor.go:
```diff
@@ -59,7 +63,56 @@ func (exec *TaskCheckStatementAdvisorExecutor) Run(ctx context.Context, server *
+
+	if payload.DbType != db.MySQL && payload.DbType != db.TiDB {
+		return []api.TaskCheckResult{}, common.Errorf(common.NotImplemented, fmt.Errorf("not implemented schema review for %s yet", payload.DbType))
+	}
```
Seems we shall put the DB check inside each rule, as the schema review rule is related to the DB type. And we also need a mapping about `SchemaReviewRuleType` - `db.Type`

rebelice replied:
Got it

d-bytebase commented on server/task_check_executor_statement_advisor.go:
```diff
@@ -59,7 +63,56 @@ func (exec *TaskCheckStatementAdvisorExecutor) Run(ctx context.Context, server *
+			payload.Statement,
+		)
+		if err != nil {
+			return []api.TaskCheckResult{}, common.Errorf(common.Internal, fmt.Errorf("failed to check statement: %w", err))
```
nit: []api.TaskCheckResult{} -> nil, as this is Go slice.

d-bytebase commented on server/task_check_scheduler.go:
```diff
@@ -333,6 +333,31 @@ func (s *TaskCheckScheduler) ScheduleCheckIfNeeded(ctx context.Context, task *ap
+			if err != nil {
+				return nil, fmt.Errorf("failed to marshal statement advise payload: %v, err: %w", task.Name, err)
+			}
+			_, err = s.server.store.CreateTaskCheckRunIfNeeded(ctx, &api.TaskCheckRunCreate{
```
nit: inline err
```
if _, err := xxx; err != nil {
   xxx
}
```

d-bytebase reviewed (dismissed):
LG otherwise

rebelice commented:
Address all comments. PTAL @tianzhou @ecmadao @d-bytebase

ecmadao reviewed (approved):
LGTM

d-bytebase commented on server/task_check_scheduler.go:
```diff
@@ -333,6 +333,30 @@ func (s *TaskCheckScheduler) ScheduleCheckIfNeeded(ctx context.Context, task *ap
+			if err != nil {
+				return nil, fmt.Errorf("failed to marshal statement advise payload: %v, err: %w", task.Name, err)
+			}
+			if _, err = s.server.store.CreateTaskCheckRunIfNeeded(ctx, &api.TaskCheckRunCreate{
```
Use if _, err := xxx instead of if _, err = xxx

rebelice commented:
@d-bytebase PTAL

tianzhou commented on api/task_check_run.go:
```diff
@@ -92,6 +92,9 @@ type TaskCheckDatabaseStatementAdvisePayload struct {
 	Collation string  `json:"collation,omitempty"`
+
+	// Schema review special fields.
+	EnvironmentID int `json:"environmentID,omitempty"`
```
I would pass the policy ID instead since the check does not necessarily need to know the environment.

Also, there is a subtle difference, if we use the environment id, what if the user later changes environment <-> policy binding after the task check is created, but before the task check is run?

It's more reasonable that the policy snapshot is created when the task check is created.

You may realize that passing the policy id still has a similar problem, what if the user later changes the policy content?

TBH, passing the entire policy content might be more reasonable from the code logic point of view. However, since the new task check payload is persisted every time we start a new run, it's a bit costly to persist that entire payload. Thus using Policy ID is a decent tradeoff.

rebelice replied:
I don't think `policy ID` should be used as an identifier here. Because `policy ID` doesn't solve the problem
> there is a subtle difference, if we use the environment id, what if the user later changes environment <-> policy binding after the task check is created, but before the task check is run?
When the environment <-> policy binding is changed, the old policy is usually deleted. In this case, the corresponding policy cannot be found through the `policy ID`.

From the user's point of view, I think there are two designs to choose from:
1. Always use the latest policy for checking. That means using the `Environment ID` as the identifier here. This design assumes that the user modifies the policy to use the updated policy. So it's reasonable to use the new policy for checks that were created before the binding was modified but run after the modification.
2. Use the policy at creation time. This design requires one more read and write operation.

I prefer the first one. Because the assumptions it brings are actually reasonable at the user level. If this assumption is not acceptable, I think only `policy snapshot `can be used instead of `policy id`.

tianzhou replied:
First, "Policy" at the creation time is the desired behavior.  Mutation after creation is problematic in the review scenario, if things change, the ideal behavior is to dismiss the previous approval if exists.

If we use "Environment ID", then we have 2 mutation points.

If we use "Policy ID", then we have 1 mutation point.

So "Policy ID" is a better solution than "Environment ID", but it's still a tradeoff compared to storing the full snapshot.

I am also fine storing the full snapshot as it removes the dependency on policy ID at all, this is desirable if we extract the "schema rule check" into a separate service (which is independent of the bytebase metadata)

e.g. we provide a SQL lint service at   api.bytebase.com/sqlcheck

rebelice replied:
OK. I think I see what you mean.

> Mutation after creation is problematic in the review scenario, if things change, the ideal behavior is to dismiss the previous approval if exists.

I would follow this behavior.
Thanks.

tianzhou replied:
As a second thought, saving the entire snapshot is still too much at this point. We want to keep our metadata as little as possible. The small metadata has several benefits:

1. Keep our own backup small.
2. Fast schema migration on our own metadata.

And because the policy snapshot is relatively big and contains duplicate info 99% of the time, so the cost is too high.

So I would go back to suggest using the "Policy ID", it's the less evil one than the "Environment ID" at least.

Sorry about the back and forth here.

rebelice commented on server/server.go:
```diff
@@ -162,6 +162,7 @@ func NewServer(logger *zap.Logger, storeInstance *store.Store, loggerLevel *zap.
 		taskCheckScheduler.Register(string(api.TaskCheckDatabaseStatementFakeAdvise), statementExecutor)
 		taskCheckScheduler.Register(string(api.TaskCheckDatabaseStatementSyntax), statementExecutor)
 		taskCheckScheduler.Register(string(api.TaskCheckDatabaseStatementCompatibility), statementExecutor)
+		taskCheckScheduler.Register(string(api.TaskCheckDatabaseStatementAdvise), statementExecutor)
```
Sounds good!

rebelice commented:
@tianzhou PTAL

tianzhou reviewed (changes requested):
Please check the replied comment

rebelice commented:
Fixed all comments. PTAL @tianzhou

tianzhou reviewed (approved):
LGTM
